stdlib/public/core/DoubleWide.swift

```diff
@@ -0,0 +1,799 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A fixed-width integer that has twice the bit width of its base type.
+///
+/// You can use the `DoubleWidth` type to continue calculations with the result
+/// of a full width arithmetic operation. Normally, when you perform a full
+/// width operation, the result is a tuple of the high and low parts of the
+/// result.
+///
+///     let a = 2241543570477705381
+///     let b = 186319822866995413
+///     let c = a.multipliedFullWidth(by: b)
+///     // c == (high: 22640526660490081, low: 7959093232766896457)
+///
+/// The tuple `c` can't be used in any further comparisons or calculations. To
+/// use this value, create a `DoubleWidth` instance from the result. You can
+/// use the `DoubleWidth` instance in the same way that you would use any other
+/// integer type.
+///
+///     let d = DoubleWidth(a.multipliedFullWidth(by: b))
+///     // d == 417644001000058515200174966092417353
+///
+///     // Check the calculation:
+///     print(d / DoubleWidth(a) == b)
+///     // Prints "true"
+///
+///     if d > Int.max {
+///         print("Too big to be an 'Int'!")
+///     } else {
+///         print("Small enough to fit in an 'Int'")
+///     }
+///     // Prints "Too big to be an 'Int'!"
+///
+/// The `DoubleWidth` type is not intended as a replacement for a variable-width
+/// integer type. Nesting `DoubleWidth` instances, in particular, may result in
+/// undesirable performance.
+@frozen
+@usableFromInline
+struct _DoubleWidth<Base: FixedWidthInteger> {
```

Comment:
Ugh, I don't think we want to add `_DoubleWidth` to the stdlib's ABI -- can we hide this behind a resiliency barrier? (I.e., remove the `@usableFromInline` attribute.)

Reply:
Or rename `_DoubleWidth` to `Duration`, and remove the APIs and conformances that aren't needed?

Reply:
Yep -- it is convenient to have a little placeholder (non-generic) struct as a stand-in for the eventual `Int128`, but preferably it shouldn't need to be ABI. (`_DoubleWidth` is way too clever  -- we can easily remove its generic argument & most of its members, as the clock use case doesn't require them.)

If we end up landing this with `_DoubleWidth` in the ABI for now, then at minimum we'll need to declare SwiftStdlib 9999 availability for this type & its members. (I wonder why the ABI checker isn't complaining about this. 🤔)

Reply:
> (I wonder why the ABI checker isn't complaining about this. 🤔)

I think the ABI checker is disabled again, possibly because [`array_cow_checks`][] isn't in the available features of CI bots.

[`array_cow_checks`]: <https://github.com/apple/swift/pull/39520#discussion_r719661983>

Reply:
> I think the ABI checker is disabled again, possibly because array_cow_checks isn't in the available features of CI bots.

Oh no -- that is extremely disconcerting. 😰

Reply:
I submitted https://github.com/apple/swift/pull/41045 to reenable the ABI checker; when that lands, the checks may uncover new issues in this PR.